### ISSUE TITLE
filter data for more than 5k items through pnpjs SPFx

### ISSUE THREAD

**Issue by sunnyEspireGit** [pnp/pnpjs#1647]
### Category
- [ ] Question
- [ ] issue

### Version

Please specify what version of the library you are using: [         "@pnp/sp": "^2.0.6"    ]

Please specify what version(s) of SharePoint you are targeting: [    SP Online     ]

*If you are not using the latest release, please update and see if the issue is resolved before submitting an issue.*

### Expected / Desired Behavior / Question
here is the situation :

I have a list with 50k items.
I have to get data from that list with a filter on status column (Status = true).
So basically there are items with status true/false.
When i will filter (Status = true) the items returned are 18k.
How will i filter data for above condition through pnpjs SPFx.

await sp.web.lists .getByTitle("listName") .items.select("ID", "Status", "Title") .filter("Status eq 'True'").get();

This is not working. :(

### Steps to Reproduce
await sp.web.lists .getByTitle("listName") .items.select("ID", "Status", "Title") .filter("Status eq 'True'").get();

**Comment by koltyakov** [comment]
PnPjs brings nothing special and no magic. If something doesn't work on the API it won't automatically work in the PnPjs.

For a large list, the applied condition has to be such that it filters down to a smaller number as a limitation. Otherwise, the threshold exception is raised by the API.

Filtering should include something else, something which logically guaranty a smaller view. This "something" should be indexed in a list.

For a server-side processing for a large list of any size you can use getAll and **NO** filtering or ordering by.

Search API also might be a helper.

Anyways, the views must be architected appropriately to support scaling list items up.

**Comment by juliemturner** [comment]
You can also consider paging for large lists... https://pnp.github.io/pnpjs/sp/items/#get-paged-items

**Comment by sunnyEspireGit** [comment]
@juliemturner : but can we use filters in what you have suggested ? like i have written below ?
await sp.web.lists .getByTitle("listName") .items.select("ID", "Status", "Title") .filter("Status eq 'True'").getPaged();

if not then getAll( ) can be used and then inside that if condition i need to apply to push the required data, but thats slow.
As i have to do search on all the data i have pushed in that array in UI.

**Comment by sunnyEspireGit** [comment]
@koltyakov : getAll( ) can be used and then inside that if condition i need to apply to push the required data, but thats slow as the filtered data will increase going forward. As i have to do search on all the data i have pushed in that array in UI.

**Comment by juliemturner** [comment]
> but can we use filters in what you have suggested ? like i have written below ?
await sp.web.lists .getByTitle("listName") .items.select("ID", "Status", "Title") .filter("Status eq 'True'").getPaged();

I can't see why not... have you tried it?

**Comment by patrick-rodgers** [comment]
getAll is rarely a good choice and we can't really encourage you to use it. We added it originally to enable certain server side processing scenarios. Def not a good choice for huge lists on the client side.

Search might be a good option to explore, as @koltyakov mentioned above.

Also, you are getting 18k items, to do what? You can't possibly intend to display them all at once. What is the main use case you are solving for?

**Comment by sunnyEspireGit** [comment]
> getAll is rarely a good choice and we can't really encourage you to use it. We added it originally to enable certain server side processing scenarios. Def not a good choice for huge lists on the client side.
> 
> Search might be a good option to explore, as @koltyakov mentioned above.
> 
> Also, you are getting 18k items, to do what? You can't possibly intend to display them all at once. What is the main use case you are solving for?

@patrick-rodgers : I am using it to display all the data on dashboard from the list, so that users can search for all the data on dashboard.

lets say users have created tickets using application, also we have migrated old data to work with this application also.

**Comment by sunnyEspireGit** [comment]
> > but can we use filters in what you have suggested ? like i have written below ?
> > await sp.web.lists .getByTitle("listName") .items.select("ID", "Status", "Title") .filter("Status eq 'True'").getPaged();
> 
> I can't see why not... have you tried it?

No i haven't tried as there was no such example on the provided in getPaged( ) using filters.

So basically what you are saying is if after apllying filter on 18k data if i get filtered data like 12k, so getpaged() will get them all in batched of 5k.

**Comment by patrick-rodgers** [comment]
A suggestion would be to update your design and have the user enter the search criteria and you then query using search to grab those items for display. Holding 18k+ items on the client isn't the best for performance, especially if you are loading everything each time they visit the page.

Otherwise give the filtered getPaged a try. We can look at adding a specific example to cover that to the docs.

**Comment by sunnyEspireGit** [comment]
> A suggestion would be to update your design and have the user enter the search criteria and you then query using search to grab those items for display. Holding 18k+ items on the client isn't the best for performance, especially if you are loading everything each time they visit the page.
> 
> Otherwise give the filtered getPaged a try. We can look at adding a specific example to cover that to the docs.

@juliemturner @patrick-rodgers : As suggested by you i have tried filter on getPaged( ) but as the filtered data is more than 5k it is also giving me threshold error. 

 ```
let resultProg = await sp.web.lists.getByTitle(GlobalConstants.lstNewProgram)
                            .items.select("ID", "Status", "Title").top(100)
                            .filter("Status eq 'Publish'")
                            .getPaged();
```

Error in console : **Error making HttpClient request in queryable [500] Internal Server Error ::> {"odata.error":{"code":"-2147024860, Microsoft.SharePoint.SPQueryThrottledException","message":{"lang":"en-US","value":"The attempted operation is prohibited because it exceeds the list view threshold."}}}**

Please suggest.

**Comment by phillipharding** [comment]
When working with large lists, if you can't solve the problem using indexed columns which align with your REST queries, you may do better using the **```RenderListDataAsStream```** method...

As already mentioned its better to try and get smaller sets of data if at all possible when your list field/schema supports querying data at more granular levels, rather than trying to deal with a very large data-set in 1 go.

With ```RenderListDataAsStream``` you have to get each page of data iteratively, the first time you call it the ```Paging``` property will be empty and each response will contain a ```NextHref``` value which you can use in the next call to ```RenderListDataAsStream``` -- repeat this process until ```NextHref``` is empty, then you have all your listitems.

_e.g._
```params["Paging"] = (response.NextHref || "").replace(/^\?/gi, "");```

```
response = await sp.web.getList(`/sites/sitename/Lists/listurlname`).renderListDataAsStream({
  AddRequiredFields: true,
  FolderServerRelativeUrl: `/sites/sitename/Lists/listurlname`,
  Paging: "",
  RenderOptions: 4103,
  ViewXml: `
<View Scope="RecursiveAll">
  <RowLimit Paged="TRUE">2000</RowLimit>
  <Query>
    <OrderBy Override="TRUE">
      <FieldRef Ascending="TRUE" Name="ID"/>
    </OrderBy>
    <Where>
      <Eq>
        <FieldRef Name="Status"/><Value Type="Text">Publish</Value>
      </Eq>
    </Where>
  </Query>
  <ViewFields>
    <FieldRef Name="ID"/>
    <FieldRef Name="Title"/>
    <FieldRef Name="Status"/>
  </ViewFields>
</View>`
});
```

**Comment by sunnyEspireGit** [comment]
> When working with large lists, if you can't solve the problem using indexed columns which align with your REST queries, you may do better using the **`RenderListDataAsStream`** method...
> 
> As already mentioned its better to try and get smaller sets of data if at all possible when your list field/schema supports querying data at more granular levels, rather than trying to deal with a very large data-set in 1 go.
> 
> With `RenderListDataAsStream` you have to get each page of data iteratively, the first time you call it the `Paging` property will be empty and each response will contain a `NextHref` value which you can use in the next call to `RenderListDataAsStream` -- repeat this process until `NextHref` is empty, then you have all your listitems.
> 
> _e.g._
> `params["Paging"] = (response.NextHref || "").replace(/^\?/gi, "");`
> 
> ```
> response = await sp.web.getList(`/sites/sitename/Lists/listurlname`).renderListDataAsStream({
>   AddRequiredFields: true,
>   FolderServerRelativeUrl: `/sites/sitename/Lists/listurlname`,
>   Paging: "",
>   RenderOptions: 4103,
>   ViewXml: `
> <View Scope="RecursiveAll">
>   <RowLimit Paged="TRUE">2000</RowLimit>
>   <Query>
>     <OrderBy Override="TRUE">
>       <FieldRef Ascending="TRUE" Name="ID"/>
>     </OrderBy>
>     <Where>
>       <Eq>
>         <FieldRef Name="Status"/><Value Type="Text">Publish</Value>
>       </Eq>
>     </Where>
>   </Query>
>   <ViewFields>
>     <FieldRef Name="ID"/>
>     <FieldRef Name="Title"/>
>     <FieldRef Name="Status"/>
>   </ViewFields>
> </View>`
> });
> ```

@phillipharding :  Thanks Phillip i have tried this, this works like getPaged() only, only difference is it is not giving error if the filtered result is more than 5k.

I got nextPage Url in  -  response.ListData.NextHref

```
private LoadAllTitles1 = async (): Promise<void> => {
    await sp.web.getList(`/sites/c3po/Lists/C3PO_ProgramTitle`).renderListDataAsStream({
                          AddRequiredFields: true,
                          FolderServerRelativeUrl: `/sites/c3po/Lists/C3PO_ProgramTitle`,
                          Paging: "",
                          RenderOptions: 4103,
                          ViewXml: 
                          `<View Scope="RecursiveAll">
                              <RowLimit Paged="TRUE">4999</RowLimit>
                              <Query>
                                <OrderBy Override="TRUE">
                                  <FieldRef Ascending="TRUE" Name="ID"/>
                                </OrderBy>
                                <Where>
                                  <Eq>
                                    <FieldRef Name="Status"/><Value Type="Text">Publish</Value>
                                  </Eq>
                                </Where>
                              </Query>
                              <ViewFields>
                                <FieldRef Name="ID"/>
                                <FieldRef Name="Title"/>
                                <FieldRef Name="Status"/>
                              </ViewFields>
                            </View>`
                        })
                        .then(async (response) => {
                          console.log("Logging response : " + response);
                          let nextUrl = ""; 
                          nextUrl = (response.ListData.NextHref || "").replace(/^\?/gi, "");

                          while (nextUrl !== '') 
                          {    
                            nextUrl = await this.CallRecursive(nextUrl);
                          }
                        })
  }

  private CallRecursive = async (nextUrl): Promise<string> => {

    return new Promise(async (resolve, reject) => {
      let response_N = await sp.web.getList(`/sites/c3po/Lists/C3PO_ProgramTitle`).renderListDataAsStream({
        AddRequiredFields: true,
        FolderServerRelativeUrl: `/sites/c3po/Lists/C3PO_ProgramTitle`,
        Paging: nextUrl,
        RenderOptions: 4103,
        ViewXml:
          `<View Scope="RecursiveAll">
            <RowLimit Paged="TRUE">4999</RowLimit>
            <Query>
              <OrderBy Override="TRUE">
                <FieldRef Ascending="TRUE" Name="ID"/>
              </OrderBy>
              <Where>
                <Eq>
                  <FieldRef Name="Status"/><Value Type="Text">Publish</Value>
                </Eq>
              </Where>
            </Query>
            <ViewFields>
              <FieldRef Name="ID"/>
              <FieldRef Name="Title"/>
              <FieldRef Name="Status"/>
            </ViewFields>
          </View>`
      });
      console.log(response_N.ListData.Row.length);
      resolve(response_N.ListData.NextHref);
  
      return response_N.ListData.NextHref;
    })
  }
```
I have tried to follow the approach suggested by you, But i am getting the same url in "NextHref" value.
Could you please help if i am missing anything ?

**Comment by phillipharding** [comment]
@sunnyEspireGit 

You have to remove the leading **?** character from the  **NextHref** value

```resolve(response_N.ListData.NextHref || "").replace(/^\?/gi, ""));```

**Comment by sunnyEspireGit** [comment]
> resolve(response_N.ListData.NextHref || "").replace(/^\?/gi, ""));

@phillipharding : What you have suggested is right, but actually my issue is different, i am getting the same (next url) for every loop. 

if you see my code in "CallRecursive" function i am passing the next url in  "Paging: nextUrl" is that correct approach  ??

**Comment by phillipharding** [comment]
it is but in your CallRecursive function you also need to resolve the NextHref removing the leading ? character

**Comment by sunnyEspireGit** [comment]
It looks like its working after making the changes you have suggested now i am getting different nextUrls - 

1. Paged=TRUE&RootFolder=%2fsites%2fc3po%2fLists%2fC3PO%5fProgramTitle&ix_Paged=TRUE&ix_ID=5001&PageFirstRow=2763&View=00000000-0000-0000-0000-000000000000"
2. ?Paged=TRUE&RootFolder=%2fsites%2fc3po%2fLists%2fC3PO%5fProgramTitle&ix_Paged=TRUE&ix_ID=10392&PageFirstRow=4925&View=00000000-0000-0000-0000-000000000000
3. ?Paged=TRUE&RootFolder=%2fsites%2fc3po%2fLists%2fC3PO%5fProgramTitle&ix_Paged=TRUE&ix_ID=15394&PageFirstRow=4974&View=00000000-0000-0000-0000-000000000000"

Thanks for all your suggestions.

@patrick-rodgers : there's one suggestion so if we are applying filters on large list and filtered data is more than 5k, then filter in getPaged() is also giving error. So @juliemturner and you have suggested to use that, that is not helpful in this situation.

While your suggestion on list architecture is worth it.

@phillipharding : Thanks, for your coding suggestion. It works perfectly how it was needed to.
I have used this kind of approach in pnp powershell to work on large list but in SPFx i wasn't aware how to do that.

This is how i have used in powershell, may be it will be helpful for someone.
 ```
$qur = "<View Scope='Recursive'><Query><OrderBy><FieldRef Name='Deal_Id' Ascending='False' /></OrderBy></Query><RowLimit>2000</RowLimit></View>"
  $listItems = Get-PnPListItem -List $DealListName  -Query $qur
```

**Comment by patrick-rodgers** [comment]
Closing this issue as answered. If you have additional questions or we did not answer your question, please open a new issue, ref this issue, and provide any additional details available. Thank you!
